### PR TITLE
Add polymorphic types for serializer GenericForeignKeys

### PR DESCRIPTION
**Pull request by glennmatthews**
# Closes: #1422
# What's Changed

This all builds atop drf-spectacular's `PolymorphicProxySerializer` class, but some details of the changes I made to make it fit our use case:

- Add an `object_type` field to all nested serializers, which simply is a string representation of the content-type of this serializer (e.g. `"dcim.device"`)
    - When using `PolymorphicProxySerializer` there needs to be some field present on each serializer that uniquely identifies which serializer it is. We didn't have that, so I added it.
    - Some of our nested serializers weren't actually inheriting from `WritableNestedSerializer` (our standard base class) so I had to change them over.
- Subclass `PolymorphicProxySerializer` to add a couple of features missing in the base class.
   - [ ] I'd like to consider offering these upstream to drf-spectacular, as they seem generally applicable.
- For each `SerializerMethodField` that we have that maps to one of several distinct nested serializers (due to use of GenericForeignKey), update the schema annotation (`@extend_schema_field`) using our custom `PolymorphicProxySerializer` to identify the possible serializers that may be seen.
   - I added a couple of helpful methods `get_all_concrete_subclasses` and `get_serializers_for_models` to make it easier to dynamically identify the possible serializers for each given field.

Before:

image

After:

image

# TODO

- [x] Explanation of Change(s)
- [x] Added change log fragment(s) (for more information see [the documentation](https://docs.nautobot.com/projects/core/en/stable/development/#creating-changelog-fragments))
- [x] Attached Screenshots, Payload Example
- n/a Unit, Integration Tests
- n/a Documentation Updates (when adding/changing features)
- n/a Example Plugin Updates (when adding/changing features)
- [x] Outline Remaining Work, Constraints from Design
   - [ ] We should consider whether to backport this to 1.5 LTM. It won't patch cleanly due to #3132, but this *should* be low risk to include in there as a bug fix.

### CODE REVIEW

File: nautobot/extras/api/nested_serializers.py
Comment (bryanculver) [inline nautobot/extras/api/nested_serializers.py:161]: Should we allow writing/editing Nested Job Log Entries?

File: nautobot/extras/api/nested_serializers.py
Comment (bryanculver) [inline nautobot/extras/api/nested_serializers.py:178]: Should we allow writing/editing Nested Job Result Entries?

File: nautobot/extras/api/nested_serializers.py
Comment (bryanculver) [inline nautobot/extras/api/nested_serializers.py:145]: Do Scheduled Job and Job Result need their nested Job Model writable?

File: nautobot/extras/api/nested_serializers.py
Comment (glennmatthews) [reply nautobot/extras/api/nested_serializers.py:161]: `WritableNestedSerializer` is not well-named - just using it doesn't automatically make a serializer or a field writable, it just provides common infra (including the new `object_type` parameter I'm adding in this PR) for serializers that *are* declared as writable. By writable it just means that *if* you're updating a foreign object reference on the base serializer, you can identify the new foreign object either by its PK or by any unique set of fields.

The only place we actually use `NestedJobLogEntrySerializer` at present is on the `api/extras/job-results/<uuid>/logs/` endpoint, which is read-only. 

Our own docs specify that nested serializers should *always* use `WritableNestedSerializer` so this is making things align better with our own docs.

Comment (glennmatthews) [reply nautobot/extras/api/nested_serializers.py:161]: Looking back at this, this change is not *strictly* necessary since `JobLogEntry` is not a full-featured model so it doesn't *have* to have an `object_type` field since it can't be used with Notes or ObjectChanges. But I think it's still a good change.

File: nautobot/extras/api/nested_serializers.py
Comment (glennmatthews) [reply nautobot/extras/api/nested_serializers.py:178]: See above comment. The only places we use NestedJobResultSerializer at present are read-only. As with JobLogEntry, `JobResult` doesn't support Notes or ObjectChanges at present so this change ends up not being *strictly* necessary but I think it's still desirable from a consistency standpoint.

File: nautobot/extras/api/nested_serializers.py
Comment (glennmatthews) [reply nautobot/extras/api/nested_serializers.py:145]: `Job` *does* support Notes and ObjectChanges, so we need an  `object_type` on NestedJobSerializer.

JobResultSerializer and ScheduledJobSerializer both specify that their associated NestedJobSerializer is `read_only=True`.

File: nautobot/extras/api/nested_serializers.py
Comment (glennmatthews) [inline nautobot/extras/api/nested_serializers.py:243]: Just calling attention to this addition in my last commit. When we specify schedule information on a JobInputSerializer, this is *not* a standard use of a nested serializer, as it's defining a new schedule for the job, *not* referencing an existing ScheduledJob instance. `WritableNestedSerializer` expects that any parameters you provide it are meant to be used to locate an existing instance and will throw a ValidationError if it doesn't find one, so for the JobInputSerializer case I had to create this separate class which is *not* a WritableNestedSerializer.

**Review by jathanism — CHANGES_REQUESTED** [review]

File: nautobot/core/api/fields.py
Comment (HanlinMiao) [inline nautobot/core/api/fields.py:123]: Just asking this for my own knowledge, what is `self.parent` here?

File: nautobot/core/api/serializers.py
Comment (HanlinMiao) [inline nautobot/core/api/serializers.py:203]: Saved one line hehe 
```suggestion
        fields = set(super().get_field_names(declared_fields, info))
        fields.add("object_type")
```

File: nautobot/dcim/api/serializers.py
Comment (glennmatthews) [reply nautobot/dcim/api/serializers.py:152]: I was thinking that `get_all_concrete_subclasses` should be useful in other future cases as well, such as perhaps the follow-on work after #3360.

File: nautobot/extras/api/nested_serializers.py
Comment (glennmatthews) [inline nautobot/extras/api/nested_serializers.py:244]:
```suggestion
    """
    Nested serializer specifically for use with `JobInputSerializer.schedule`.

    We don't use `WritableNestedSerializer` here because this is not used to look up
    an existing `ScheduledJob`, but instead used to specify parameters for creating one.
    """
```

File: nautobot/extras/api/nested_serializers.py
Comment (glennmatthews) [reply nautobot/extras/api/nested_serializers.py:243]: Will do.

**Review by HanlinMiao — APPROVED** [review]
Do we need any tests for this feature?

File: nautobot/extras/api/serializers.py
Comment (glennmatthews) [inline nautobot/extras/api/serializers.py:952]: Here and in `extras/api/views.py` and in `extras/tests/test_api.py` I'm reverting this serializer field from `scheduled_job` (change made in #3262) back to `schedule`; as discussed, this isn't actually a reference to a ScheduledJob instance, it's a schedule for the job that may or may not result in the creation of a ScheduledJob instance, so I think `schedule` is more accurate.

File: nautobot/ipam/api/serializers.py
Comment (glennmatthews) [inline nautobot/ipam/api/serializers.py:362]: Could arguably use `get_all_concrete_subclasses(BaseInterface)` here instead.

File: nautobot/dcim/api/serializers.py
Comment (glennmatthews) [reply nautobot/dcim/api/serializers.py:152]: Also there are cases where I'm using `get_serializers_for_models` separarately, such as:

```python
            serializers=lambda: get_serializers_for_models(
                FeatureQuery("config_context_owners").list_subclasses(), prefix="Nested"
            ),
```

Or were you suggesting that I keep both of these as separate functions but add a third wrapper function that combines the two?

**Comment by glennmatthews** [comment]
> Do we need any tests for this feature?

We don't have any current tests for the generated schema (other than running `nautobot-server check-schema` in the CI to make sure that it at least generates without errors). I suppose it'd be possible to write a set of tests that would render the schema to YAML then parse the YAML and look for specific data structures. Mind if I add that as a backlog item?

**Comment by HanlinMiao** [comment]
> > Do we need any tests for this feature?
> 
> We don't have any current tests for the generated schema (other than running `nautobot-server check-schema` in the CI to make sure that it at least generates without errors). I suppose it'd be possible to write a set of tests that would render the schema to YAML then parse the YAML and look for specific data structures. Mind if I add that as a backlog item?

Yeah, I think the test is out of scope for this PR at least.

**Comment by glennmatthews** [comment]
On hold until I can verify that this works correctly with `oapi-codegen`.

**Comment by glennmatthews** [comment]
> > > Do we need any tests for this feature?
> > 
> > 
> > We don't have any current tests for the generated schema (other than running `nautobot-server check-schema` in the CI to make sure that it at least generates without errors). I suppose it'd be possible to write a set of tests that would render the schema to YAML then parse the YAML and look for specific data structures. Mind if I add that as a backlog item?
> 
> Yeah, I think the test is out of scope for this PR at least.

I added a basic test framework and a couple of sample tests in f08809d.

File: nautobot/dcim/api/serializers.py
Comment (jathanism) [reply nautobot/dcim/api/serializers.py]: Oh yeah, I was thinking to leave the other two helpers in place, but just having this other one for the common case to wrap them if the decomposition isn't needed. Sorry fit hat wasn't clear.

A third helper to wrap the two helpers for the common case.

File: nautobot/dcim/tests/test_schema.py
Comment (jathanism) [inline nautobot/dcim/tests/test_schema.py:6]: I want to be clear in that I think this is awesome.

What I don't like is how there is so much hard-coding of object names and nested fields relationships. I wish there were a way we could abstract this somehow.

If we were, for example, to manually add schema tests across the board, it would be incredibly tedious.

I know you added these as sample tests and it seems reasonable that this lays the foundation for implementing generic tests based on expectations provided by the way things are marked up using schema decorators?

**Comment by bryanculver** [comment]
What requires this to go to `next` vs `develop`? If I understand the PR correctly it just makes the schema reflect more accurately what you've always had to send to the API but previously was described as a generic JSON dictionary. Is it quasi-breaking in that it will do better serializer enforcement? Or because of:
> Add an `object_type` field to all nested serializers, which simply is a string representation of the content-type of this serializer (e.g. "dcim.device")


will this reject previous calls missing the `object_type` in a request payload?

**Review by bryanculver — APPROVED** [review]

**Comment by bryanculver** [comment]
> > What requires this to go to `next` vs `develop`?
> 
> Nothing specific, I just started working on this under `next` initially as I wasn't initially sure of the scope and impact. As I noted in the PR description:
> 
> > _We should consider whether to backport this to 1.5 LTM. It won't patch cleanly due to #3132, but this should be low risk to include in there as a bug fix._
> 
> I'll add that the later-added changes in this PR to the `JobInputSerializer` are specific to `next` and wouldn't apply to `develop` at all. So it's not a clean port over to `develop` for a number of reasons, but it _is_ a bug fix and _could_ be committed back to `develop` for the 1.5 LTM if we think the effort is justified.

Missed the last line in the original description. I think it's fine to leave it to `next` unless there is a significant push to backport it.

File: nautobot/dcim/tests/test_schema.py
Comment (glennmatthews) [reply nautobot/dcim/tests/test_schema.py:6]: As discussed, I'll take a quick stab at refining this a bit to reduce the repetition.

Comment (glennmatthews) [reply nautobot/dcim/tests/test_schema.py:6]: Or spend a couple hours on it, whichever. 😅 Please take a look at the updated tests.

File: nautobot/core/api/serializers.py
Comment (glennmatthews) [inline nautobot/core/api/serializers.py:304]: I contributed the serializers-as-callable feature upstream as https://github.com/tfranzel/drf-spectacular/pull/953, so once there's a new release we can remove some of this code.

Comment (bryanculver) [reply nautobot/core/api/serializers.py:304]: 😎 https://github.com/tfranzel/drf-spectacular/releases/tag/0.26.0

Comment (bryanculver) [reply nautobot/core/api/serializers.py]: I have removed these methods and updated to `0.26` with 1b10237.

Since `allow_null` is not upstream I figured it would be preferred to be consistent to still use our overloaded class than split hairs between where we used ours vs what could use the default from DRF-Spectacular (example `origin` on `CablePathSerializer`).

Comment (glennmatthews) [reply nautobot/core/api/serializers.py]: Yup, good stuff. I think the `allow_null` here doesn't quite work across the board, or at least that's what I was seeing when I put a little time last week into trying to port it upstream as well. (IIRC it works for our case where we're using a polymorphic serializer for a specific field on an API endpoint, but it didn't seem to work as expected for the upstream test case that has an entire API endpoint that's polymorphic.) So for the moment keeping our local `allow_null` enhancement here makes sense.

**Review by jathanism — APPROVED** [review]

**Pull request merged**
